File: frameworks/juniper/tests/test_helpers.rs
use uuid::Uuid;
use chrono::Utc;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

/// TestFactory provides helper methods for creating test data
pub struct TestFactory {
    users: Arc<Mutex<HashMap<String, TestUser>>>,
    posts: Arc<Mutex<HashMap<String, TestPost>>>,
}

#[derive(Clone, Debug)]
pub struct TestUser {
    pub id: String,
    pub pk_user: i32,
    pub username: String,
    pub full_name: String,
    pub bio: Option<String>,
    pub created_at: String,
}

#[derive(Clone, Debug)]
pub struct TestPost {
    pub id: String,
    pub pk_post: i32,
    pub title: String,
    pub content: String,
    pub fk_author: i32,
    pub created_at: String,
}

#[derive(Clone, Debug)]
pub struct TestComment {
    pub id: String,
    pub pk_comment: i32,
    pub content: String,
    pub fk_post: i32,
    pub fk_author: i32,
    pub created_at: String,
}

impl TestFactory {
    /// Create a new test factory instance
    pub fn new() -> Self {
        TestFactory {
            users: Arc::new(Mutex::new(HashMap::new())),
            posts: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Create a test user
    pub fn create_test_user(
        &self,
        username: &str,
        _email: &str,
        full_name: &str,
        bio: &str,
    ) -> TestUser {
        let id = Uuid::new_v4().to_string();
        let pk_user = self.users.lock().unwrap().len() as i32 + 1;
        let bio = if bio.is_empty() { None } else { Some(bio.to_string()) };

        let user = TestUser {
            id: id.clone(),
            pk_user,
            username: username.to_string(),
            full_name: full_name.to_string(),
            bio,
            created_at: Utc::now().to_rfc3339(),
        };

        self.users.lock().unwrap().insert(id, user.clone());
        user
    }

    /// Create a test post
    pub fn create_test_post(&self, author_id: &str, title: &str, content: &str) -> TestPost {
        let id = Uuid::new_v4().to_string();
        let pk_post = self.posts.lock().unwrap().len() as i32 + 1;
        let fk_author = self
            .users
            .lock()
            .unwrap()
            .values()
            .find(|u| u.id == author_id)
            .map(|u| u.pk_user)
            .unwrap_or(1);

        let post = TestPost {
            id: id.clone(),
            pk_post,
            title: title.to_string(),
            content: content.to_string(),
            fk_author,
            created_at: Utc::now().to_rfc3339(),
        };

        self.posts.lock().unwrap().insert(id, post.clone());
        post
    }

    /// Create a test comment
    pub fn create_test_comment(&self, author_id: &str, post_id: &str, content: &str) -> TestComment {
        let id = Uuid::new_v4().to_string();
        let pk_comment = 1; // Simplified for tests
        let fk_author = self
            .users
            .lock()
            .unwrap()
            .values()
            .find(|u| u.id == author_id)
            .map(|u| u.pk_user)
            .unwrap_or(1);
        let fk_post = self
            .posts
            .lock()
            .unwrap()
            .values()
            .find(|p| p.id == post_id)
            .map(|p| p.pk_post)
            .unwrap_or(1);

        TestComment {
            id: id.clone(),
            pk_comment,
            content: content.to_string(),
            fk_post,
            fk_author,
            created_at: Utc::now().to_rfc3339(),
        }
    }

    /// Get a user by ID
    pub fn get_user(&self, id: &str) -> Option<TestUser> {
        self.users.lock().unwrap().get(id).cloned()
    }

    /// Get a post by ID
    pub fn get_post(&self, id: &str) -> Option<TestPost> {
        self.posts.lock().unwrap().get(id).cloned()
    }

    /// Get all users
    pub fn get_all_users(&self) -> Vec<TestUser> {
        self.users
            .lock()
            .unwrap()
            .values()
            .cloned()
            .collect()
    }

    /// Get all posts
    pub fn get_all_posts(&self) -> Vec<TestPost> {
        self.posts
            .lock()
            .unwrap()
            .values()
            .cloned()
            .collect()
    }

    /// Get user count
    pub fn user_count(&self) -> usize {
        self.users.lock().unwrap().len()
    }

    /// Get post count
    pub fn post_count(&self) -> usize {
        self.posts.lock().unwrap().len()
    }

    /// Reset all test data
    pub fn reset(&self) {
        self.users.lock().unwrap().clear();
        self.posts.lock().unwrap().clear();
    }
}

impl Default for TestFactory {
    fn default() -> Self {
        Self::new()
    }
}

/// ValidationHelper provides common validation assertions
pub struct ValidationHelper;

impl ValidationHelper {
    /// Validate UUID format
    pub fn assert_uuid(value: &str) -> Result<(), String> {
        match Uuid::parse_str(value) {
            Ok(_) => Ok(()),
            Err(e) => Err(format!("Invalid UUID: {} - {}", value, e)),
        }
    }

    /// Validate string is not empty
    pub fn assert_not_empty(value: &str, name: &str) -> Result<(), String> {
        if value.is_empty() {
            Err(format!("{} should not be empty", name))
        } else {
            Ok(())
        }
    }

    /// Validate option is not none
    pub fn assert_is_some<T>(value: &Option<T>, name: &str) -> Result<(), String> {
        if value.is_none() {
            Err(format!("{} should not be None", name))
        } else {
            Ok(())
        }
    }

    /// Validate option is none
    pub fn assert_is_none<T>(value: &Option<T>, name: &str) -> Result<(), String> {
        if value.is_some() {
            Err(format!("{} should be None", name))
        } else {
            Ok(())
        }
    }

    /// Validate equality
    pub fn assert_equal<T: PartialEq + std::fmt::Debug>(
        expected: &T,
        actual: &T,
        name: &str,
    ) -> Result<(), String> {
        if expected != actual {
            Err(format!(
                "{} mismatch: expected {:?}, got {:?}",
                name, expected, actual
            ))
        } else {
            Ok(())
        }
    }
}

/// GraphQLTestHelper provides GraphQL-specific testing utilities
pub struct GraphQLTestHelper;

impl GraphQLTestHelper {
    /// Check if query has errors
    pub fn has_errors(response: &str) -> bool {
        response.contains("\"errors\"") || response.contains("error")
    }

    /// Extract error message
    pub fn extract_error_message(response: &str) -> Option<String> {
        if let Some(start) = response.find("\"message\"") {
            if let Some(quote_start) = response[start + 10..].find('\"') {
                if let Some(quote_end) = response[start + 10 + quote_start + 1..].find('\"') {
                    let msg_start = start + 10 + quote_start + 1;
                    let msg_end = msg_start + quote_end;
                    return Some(response[msg_start..msg_end].to_string());
                }
            }
        }
        None
    }
}

/// DataGenerator provides test data generation utilities
pub struct DataGenerator;

impl DataGenerator {
    /// Generate a string of specified length
    pub fn generate_long_string(length: usize) -> String {
        let mut result = String::with_capacity(length);
        for i in 0..length {
            result.push_str(&(i % 10).to_string());
        }
        result
    }

    /// Generate unique strings
    pub fn generate_unique_strings(count: usize) -> Vec<String> {
        (0..count)
            .map(|_| Uuid::new_v4().to_string())
            .collect()
    }

    /// Generate test users
    pub fn generate_users(factory: &TestFactory, count: usize) -> Vec<TestUser> {
        (0..count)
            .map(|i| {
                factory.create_test_user(
                    &format!("user{}", i),
                    &format!("user{}@example.com", i),
                    &format!("User {}", i),
                    "",
                )
            })
            .collect()
    }
}

File: frameworks/go-gqlgen/test_helpers.go
package main

import (
	"context"
	"fmt"
	"testing"

	"github.com/google/uuid"
	"github.com/benchmark/go-gqlgen/graph/model"
)

// TestFactory provides helper methods for creating test data
type TestFactory struct {
	t     *testing.T
	users map[string]*model.User
	posts map[string]*model.Post
}

// NewTestFactory creates a new test factory instance
func NewTestFactory(t *testing.T) *TestFactory {
	return &TestFactory{
		t:     t,
		users: make(map[string]*model.User),
		posts: make(map[string]*model.Post),
	}
}

// CreateTestUser creates a test user with provided details
func (tf *TestFactory) CreateTestUser(username, email, fullName, bio string) *model.User {
	id := uuid.New().String()
	user := &model.User{
		ID:       id,
		Username: username,
		FullName: &fullName,
		Bio:      &bio,
	}

	if fullName == "" {
		user.FullName = nil
	}
	if bio == "" {
		user.Bio = nil
	}

	tf.users[id] = user
	return user
}

// CreateTestPost creates a test post with provided details
func (tf *TestFactory) CreateTestPost(authorID, title, content string) *model.Post {
	id := uuid.New().String()
	post := &model.Post{
		ID:      id,
		Title:   title,
		Content: &content,
		Author: &model.User{
			ID: authorID,
		},
	}

	if content == "" {
		post.Content = nil
	}

	tf.posts[id] = post
	return post
}

// CreateTestComment creates a test comment
func (tf *TestFactory) CreateTestComment(authorID, postID, content string) *model.Comment {
	id := uuid.New().String()
	comment := &model.Comment{
		ID:      id,
		Content: content,
		Author: &model.User{
			ID: authorID,
		},
		Post: &model.Post{
			ID: postID,
		},
	}
	return comment
}

// GetUser retrieves a user by ID from factory cache
func (tf *TestFactory) GetUser(id string) *model.User {
	return tf.users[id]
}

// GetPost retrieves a post by ID from factory cache
func (tf *TestFactory) GetPost(id string) *model.Post {
	return tf.posts[id]
}

// Reset clears all test data
func (tf *TestFactory) Reset() {
	tf.users = make(map[string]*model.User)
	tf.posts = make(map[string]*model.Post)
}

// ValidationHelper provides common validation assertions
type ValidationHelper struct {
	t *testing.T
}

// NewValidationHelper creates validation helper
func NewValidationHelper(t *testing.T) *ValidationHelper {
	return &ValidationHelper{t: t}
}

// AssertUUID validates UUID format
func (vh *ValidationHelper) AssertUUID(value string) {
	_, err := uuid.Parse(value)
	if err != nil {
		vh.t.Errorf("Invalid UUID: %s - %v", value, err)
	}
}

// AssertNotNil asserts value is not nil
func (vh *ValidationHelper) AssertNotNil(value interface{}, name string) {
	if value == nil {
		vh.t.Errorf("%s should not be nil", name)
	}
}

// AssertNil asserts value is nil
func (vh *ValidationHelper) AssertNil(value interface{}, name string) {
	if value != nil {
		vh.t.Errorf("%s should be nil, got %v", name, value)
	}
}

// AssertEqual asserts two values are equal
func (vh *ValidationHelper) AssertEqual(expected, actual interface{}, name string) {
	if expected != actual {
		vh.t.Errorf("%s mismatch: expected %v, got %v", name, expected, actual)
	}
}

// AssertNotEqual asserts two values are not equal
func (vh *ValidationHelper) AssertNotEqual(notExpected, actual interface{}, name string) {
	if notExpected == actual {
		vh.t.Errorf("%s should not equal %v", name, actual)
	}
}

// AssertStringContains checks if string contains substring
func (vh *ValidationHelper) AssertStringContains(str, substring, name string) {
	if len(str) == 0 || len(substring) == 0 {
		return
	}
	// Simple string contains check
	found := false
	for i := 0; i <= len(str)-len(substring); i++ {
		if str[i:i+len(substring)] == substring {
			found = true
			break
		}
	}
	if !found {
		vh.t.Errorf("%s should contain '%s', got '%s'", name, substring, str)
	}
}

// GraphQLTestHelper provides GraphQL-specific testing utilities
type GraphQLTestHelper struct {
	t *testing.T
}

// NewGraphQLTestHelper creates GraphQL test helper
func NewGraphQLTestHelper(t *testing.T) *GraphQLTestHelper {
	return &GraphQLTestHelper{t: t}
}

// AssertNoErrors checks GraphQL response has no errors
func (gth *GraphQLTestHelper) AssertNoErrors(errors []error) {
	if len(errors) > 0 {
		gth.t.Errorf("Expected no GraphQL errors, got %d", len(errors))
		for _, err := range errors {
			gth.t.Logf("  - %v", err)
		}
	}
}

// AssertErrorCount validates error count
func (gth *GraphQLTestHelper) AssertErrorCount(expectedCount int, actualErrors []error) {
	if len(actualErrors) != expectedCount {
		gth.t.Errorf("Expected %d errors, got %d", expectedCount, len(actualErrors))
	}
}

// ContextWithFactory creates context with test factory
func ContextWithFactory(ctx context.Context, factory *TestFactory) context.Context {
	return context.WithValue(ctx, "factory", factory)
}

// DataGenerator provides test data generation utilities
type DataGenerator struct {
	t *testing.T
}

// NewDataGenerator creates data generator
func NewDataGenerator(t *testing.T) *DataGenerator {
	return &DataGenerator{t: t}
}

// GenerateLongString creates a string of specified length
func (dg *DataGenerator) GenerateLongString(length int) string {
	result := ""
	for i := 0; i < length; i++ {
		result += fmt.Sprintf("%d", i%10)
	}
	return result
}

// GenerateUniqueStrings creates N unique strings
func (dg *DataGenerator) GenerateUniqueStrings(count int) []string {
	result := make([]string, count)
	for i := 0; i < count; i++ {
		result[i] = uuid.New().String()
	}
	return result
}
